File: software/mobile/ios/SVCMobile/Sources/BLE/MockBLETransport.swift
import Foundation

final class MockBLETransport: BLETransport {
    var onDevicesChanged: (([DiscoveredDevice]) -> Void)?
    var onConnectionChanged: ((ConnectionState) -> Void)?
    private(set) var writes: [(Data, UUID)] = []

    let device = DiscoveredDevice(
        id: UUID(uuidString: "D78F3100-2AE5-4B2B-9294-0F299E60AA01")!,
        name: "SVC-MOCK-001",
        signalStrength: -42
    )

    func startScan() {
        onConnectionChanged?(.scanning)
        onDevicesChanged?([device])
    }

    func stopScan() {}

    func connect(to id: UUID) {
        onConnectionChanged?(id == device.id ? .connected : .disconnected)
    }

    func write(_ data: Data, characteristic: UUID) {
        writes.append((data, characteristic))
    }
}

File: software/mobile/android/core-ble/src/main/kotlin/com/avlyubimov/svc/core/ble/MockBleTransport.kt
package com.avlyubimov.svc.core.ble

import com.avlyubimov.svc.core.model.ConnectionState
import com.avlyubimov.svc.core.model.DiscoveredDevice
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class MockBleTransport : BleTransport {
    private val mockDevice = DiscoveredDevice(
        id = "D78F3100-2AE5-4B2B-9294-0F299E60AA01",
        name = "SVC-MOCK-001",
        signalStrength = -44,
    )
    private val mutableState = MutableStateFlow(ConnectionState.DISCONNECTED)
    private val mutableDevices = MutableStateFlow<List<DiscoveredDevice>>(emptyList())
    val writes = mutableListOf<Pair<String, ByteArray>>()

    override val connectionState: StateFlow<ConnectionState> = mutableState.asStateFlow()
    override val devices: StateFlow<List<DiscoveredDevice>> = mutableDevices.asStateFlow()

    override fun startScan() {
        mutableState.value = ConnectionState.SCANNING
        mutableDevices.value = listOf(mockDevice)
    }

    override fun stopScan() = Unit

    override fun connect(deviceId: String) {
        mutableState.value = if (deviceId == mockDevice.id) {
            ConnectionState.CONNECTED
        } else {
            ConnectionState.DISCONNECTED
        }
    }

    override fun write(characteristicUuid: String, data: ByteArray) {
        writes += characteristicUuid to data.copyOf()
    }
}
